### PR TITLE
Added animated numbers component

### LINKED ISSUE #4: Implement price change animation for the `AuctionPrice`
**Goal**
When the price drops, it counts down dynamically.

**Context**
When the new price for an auction is fetched, we currently just override the old price with the new one, without any animation. Now, we would like to have an animation that visualizes this price change a bit better. 
See https://makerburn.com/#/liquidations for inspiration.
<details>
  <summary>Switch between Quarter, Month, Week, Day to see the animation</summary>
  
https://user-images.githubusercontent.com/36764271/140878366-6e3e6ba9-2f4f-4783-ab0b-0d5bfca4198c.mp4

</details>

**Tasks**
- [ ] Research a library which can be used for this (or propose another solution)
  - [ ] If there are multiple, please bring up Pro & Contra
- [ ] Ping @florianstoecker for quick discussion on Slack
- [ ] Implement this component (storybook only)

### PR DESCRIPTION
**Pull request by benediktkaiser**
closes #4
 
Preview:

https://user-images.githubusercontent.com/30908158/150323214-449e0428-adba-42eb-a280-9d4ebb1d9331.mov

There is one issue I ran into that I personally do not like. The numbers will animate even on initial load. Meaning that I open the Auction on the side panel and I still get the animation (in this case from 0 to their initial value (see video below where I open the `PriceDropComponent` for the first time and it still animates)). 
I looked into a fix and I would be able to change this behaviour, but I would need to fork the repo and make republish it under `sidestream-tech` or make a PR to their main repo (might take a while to get merged as the package isn't the most active anymore).

Alternatly we could say that it does not bother us that much and then I wouldn't need to do anything about this. However I am leaving this call up to @LukSteib and @valiafetisov. Some opinions would be great!

https://user-images.githubusercontent.com/30908158/150322868-14b694f4-9d75-4ca2-824d-0a80ccd32f56.mov

### CODE REVIEW

**Comment by LukSteib** [comment]
> I looked into a fix and I would be able to change this behaviour, but I would need to fork the repo and make republish it under sidestream-tech or make a PR to their main repo

This procedure to me sounds to involved given the value this feature provides (is basically cosmetics only). 
Question is rather: Does the animation in its current state (with the downside you mentioned above) improve UX or add unnecessary noise compared to status quo? 

In my opinion the animation as is (shown in video 1) is fine and makes the user a little more aware of which figures actually change based on the price drop. The downside (video 2) of having this initial animation from 0 up is neglectable.

**Comment by benediktkaiser** [comment]
> user a little more aware of which figures actually change based on the price drop

I 100% agree with this. It helps pull the users attention to it which is really good. However, I have to admit that it is a lot of sudden movement, which is something we don't have anywhere else on our otherwise very clean UI. Maybe we want to only animate one or two of the numbers instead of everything? I really am not too sure what to do in this case.

It would interest me what @valiafetisov has to say about it, as I know he is very against too much noise.

**Comment by valiafetisov** [comment]
I think it's fine that numbers animate on load. 

Disabling this behaviour can be a small improvement issue: I agree it is cleaner, but I don't see too much downside. Regarding forking: until we have a strict policy, it's fine to make a private fork in order to submit a PR. Republishing a fork is not a good practice 🙂

**Comment by benediktkaiser** [comment]
I had to do some catching of invalid numbers to fix the test. (Such as when a BigNumber returns `NaN`). 

I also added the ability to deactivate the animation. Currently I did this for all numbers that won't change after each price drop and all numbers in the main AuctionTable, as I believe that if a user were to be waiting for a auction price to drop, they would have the auction selected. Therefore I believed that reducing the movement on the start page would be a good idea.

If you have any other places I should deactivate/reactivate the animation, please let me know!

**Comment by valiafetisov** [comment]
> Currently I did this for all numbers that won't change after each price drop and all numbers in the main AuctionTable

Why do you need to deactivate animation for the numbers that wouldn't change?
Are they animated even if they won't change?

**Comment by benediktkaiser** [comment]
> Are they animated even if they won't change?

On the initial opening of the panel or loading of the page they are.

**Comment by valiafetisov** [comment]
> On the initial opening

But you deactivated them completely, not only on the initial render, correct?

**Comment by benediktkaiser** [comment]
> But you deactivated them completely, not only on the initial render, correct?

Yes, as they won't change afterwards anyway. For example the Auction Amount (The amount of Collateral in a vault) won't change right?

**Comment by valiafetisov** [comment]
> For example the Auction Amount (The amount of Collateral in a vault) won't change right?

This particular number can change in case of partial bidding. We refetch this number and therefore it can change. As stated above, initial animation of the numbers is a minor inconvenience that can be addressed in a separate issue (which you can create). But I don't think that applying this workaround is good idea

**Comment by benediktkaiser** [comment]
> This particular number can change in case of partial bidding. We refetch this number and therefore it can change. As stated above, initial animation of the numbers is a minor inconvenience that can be addressed in a separate issue (which you can create). But I don't think that applying this workaround is good idea

Sounds good! Ill remove the ability to deactivate the animation completely then.

**Comment by valiafetisov** [comment]
Let's remove it completely and address initial animation in a separate issue. For all other numbers I think it would be beneficial to see them animated as soon as they change

File: frontend/components/utils/FormatCurrency.vue [hunk 1]
```diff
@@ -1,19 +1,24 @@
 <template>
-    <span
-        >{{ sign }}{{ formattedNumber }}<span class="uppercase"> {{ currency }}</span></span
-    >
+    <span>
+        {{ sign }}
+        <animated-number v-if="isValidNumber" :value="value" :decimal-places="decimals" />
```
Comment (valiafetisov) [inline frontend/components/utils/FormatCurrency.vue:4]: Since `{{ sign }}` and `<animated-number` are on different lines, when rendered, they will have space in-between, while they had none before

File: frontend/components/utils/FormatCurrency.vue [hunk 2]
```diff
@@ -44,6 +43,15 @@ export default Vue.extend({
             }
             return this.value > 0 ? '+' : '';
         },
+        isValidNumber(): boolean {
+            if (BigNumber.isBigNumber(this.value)) {
+                if (this.value.toString() === 'NaN') {
+                    return false;
+                }
+            }
```
Comment (valiafetisov) [inline frontend/components/utils/FormatCurrency.vue:51]: Why not use existing method?

```suggestion
            if (BigNumber.isBigNumber(this.value) && this.value.isNaN()) {
                return false;
            }
```

https://mikemcl.github.io/bignumber.js/#isNaN

Also, where do you actually test this edge case?

File: frontend/components/utils/AnimatedNumber.vue [hunk 3]
```diff
@@ -0,0 +1,38 @@
+<template>
+    <animated-number
+        v-if="value"
+        :value="value.toFixed(decimalPlaces)"
+        :format-value="formatValue"
```
[the remark this answers is not in the document]
Comment (benediktkaiser) [reply frontend/components/utils/AnimatedNumber.vue:5]: `formatValue` is used for the format of the "transition" numbers. This is to make sure that when it counts up from 0 to 100 the numbers correspond with the `formatValue` style. 


> * provide value as string

Do you want this component to only accept strings or should I just replace `.toFixed` with `.toString()`?

File: frontend/components/utils/FormatCurrency.vue (hunk 1 above)
Comment (benediktkaiser) [reply frontend/components/utils/FormatCurrency.vue:4]: Fixed this!

File: frontend/components/utils/FormatCurrency.vue [hunk 4]
```diff
@@ -44,6 +43,15 @@ export default Vue.extend({
...
+
+            return !!this.value;
```
[the remark this answers is not in the document]
Comment (benediktkaiser) [reply frontend/components/utils/FormatCurrency.vue:53]: Added a check plus a story for this.

File: frontend/components/utils/FormatCurrency.vue (hunk 2 above)
Comment (benediktkaiser) [reply frontend/components/utils/FormatCurrency.vue:51]: > Also, where do you actually test this edge case?

This was an issue that arose in `generateFakeAuction` resulting in failing tests and errors in the storyboard. It was never an issue before this, as BigNumber just displayed the missing BigNumber as `NaN`, however this new component threw an error when this kind of BigNumber was given. That is why I needed to add this check. 

I also added a Story, just to be aware of this edge case.

File: frontend/components/utils/AnimatedNumber.vue (hunk 3 above)
[the remark this answers is not in the document]
Comment (valiafetisov) [reply frontend/components/utils/AnimatedNumber.vue]: If this component (`animated-number-vue`) accepts (and uses) `formatValue` function that expects number/bignumber, then where does it get number/bignumber if you only give it a string?

> Do you want this component to only accept strings or should I just replace `.toFixed` with `.toString()`?

What's wrong with providing value as number as suggested above?

[the remark this answers is not in the document]
Comment (benediktkaiser) [reply frontend/components/utils/AnimatedNumber.vue]: > What's wrong with providing value as `number` without any conversion as suggested above? Since the conversion should happen via `formatValue`

It wouldn't be an issue with just numbers, however we run into the issue that the component does not support BigNumbers. Meaning I can:

a.) check if the value is a BigNumber, if yes add `.toNumber()` (this function however does not work with normal numbers or string)
b.) convert numbers and BigNumbers to strings in order to not have to check if it is a BigNumber as both BigNumbers and numbers can be converted to string with either `.toString()` or `.toFixed()`.

File: frontend/components/utils/FormatCurrency.vue (hunk 4 above)
[the remark this answers is not in the document]
Comment (valiafetisov) [reply frontend/components/utils/FormatCurrency.vue:51]: I don't see the changes here. `isValidNumber` will still be false for `0`

File: frontend/components/utils/AnimatedNumber.vue (hunk 3 above)
[the remark this answers is not in the document]
Comment (valiafetisov) [reply frontend/components/utils/AnimatedNumber.vue]: Ok, then, this component somehow converts this value back into number in order to pass to the `fromatValue` again 😅 

Please go with `b.)` but please use the same `formatValue` method for conversion

File: frontend/components/utils/FormatCurrency.vue (hunk 4 above)
[the remark this answers is not in the document]
Comment (benediktkaiser) [reply frontend/components/utils/FormatCurrency.vue]: Done now!

File: frontend/components/utils/AnimatedNumber.vue (hunk 3 above)
[the remark this answers is not in the document]
Comment (benediktkaiser) [reply frontend/components/utils/AnimatedNumber.vue]: Sounds good!

File: frontend/components/utils/AnimatedNumber.vue [hunk 5]
```diff
@@ -1,7 +1,7 @@
 <template>
     <animated-number
         v-if="value || value === 0"
```
Comment (valiafetisov) [inline frontend/components/utils/AnimatedNumber.vue:3]: Since value can be a big number 0, isn't it better to check it against undefined and null, same as in the other file?

Comment (benediktkaiser) [reply frontend/components/utils/AnimatedNumber.vue]: I added the same check on this component.

File: frontend/components/utils/FormatCurrency.vue [hunk 6]
```diff
@@ -47,8 +47,7 @@ export default Vue.extend({
             if (BigNumber.isBigNumber(this.value) && this.value.isNaN()) {
                 return false;
             }
-
-            return !!this.value;
+            return !(this.value === undefined || null);
```
[the remark this answers is not in the document]
Comment (benediktkaiser) [reply frontend/components/utils/FormatCurrency.vue]: Okay! I just checked if for `undefined `.

**Comment by benediktkaiser** [comment]
I had to add a null check again, however I did it a bit differently this time.

**Review by valiafetisov — APPROVED** [review]

**Review by florianstoecker — APPROVED** [review]
Really like the feature!! :) 

But I agree that the initial loading animation can be deactivated for some of the numbers.

**Comment by valiafetisov** [comment]
> But I agree that the initial loading animation can be deactivated

@florianstoecker, @benediktkaiser let's open an issue about that and potentially open a PR with a new prop to the library.
